Framework/Geometry/src/Instrument/GridDetector.cpp

```diff
@@ -0,0 +1,939 @@
+#include "MantidGeometry/Instrument/GridDetector.h"
+#include "MantidGeometry/Instrument/ComponentInfo.h"
+#include "MantidGeometry/Instrument/ComponentVisitor.h"
+#include "MantidGeometry/Instrument/Detector.h"
+#include "MantidGeometry/Instrument/GridDetectorPixel.h"
+#include "MantidGeometry/Objects/BoundingBox.h"
+#include "MantidGeometry/Objects/CSGObject.h"
+#include "MantidGeometry/Objects/IObject.h"
+#include "MantidGeometry/Objects/ShapeFactory.h"
+#include "MantidGeometry/Objects/Track.h"
+#include "MantidGeometry/Rendering/GeometryHandler.h"
+#include "MantidKernel/Exception.h"
+#include "MantidKernel/Material.h"
+#include "MantidKernel/Matrix.h"
+#include <algorithm>
+#include <boost/make_shared.hpp>
+#include <boost/regex.hpp>
+#include <ostream>
+#include <stdexcept>
+
+namespace Mantid {
+namespace Geometry {
+
+using Kernel::Matrix;
+using Kernel::V3D;
+
+/** Empty constructor
+ */
+GridDetector::GridDetector()
+    : CompAssembly(), IObjComponent(nullptr), m_gridBase(nullptr),
+      m_minDetId(0), m_maxDetId(0) {
+
+  init();
+  setGeometryHandler(new GeometryHandler(this));
+}
+
+/** Constructor for a parametrized GridDetector
+ * @param base: the base (un-parametrized) GridDetector
+ * @param map: pointer to the ParameterMap
+ * */
+GridDetector::GridDetector(const GridDetector *base, const ParameterMap *map)
+    : CompAssembly(base, map), IObjComponent(nullptr), m_gridBase(base),
+      m_minDetId(0), m_maxDetId(0) {
+  init();
+  setGeometryHandler(new GeometryHandler(this));
+}
+
+/** Valued constructor
+ *  @param n :: name of the assembly
+ *  @param reference :: the parent Component
+ *
+ * 	If the reference is an object of class Component,
+ *  normal parenting apply. If the reference object is
+ *  an assembly itself, then in addition to parenting
+ *  this is registered as a children of reference.
+ */
+GridDetector::GridDetector(const std::string &n, IComponent *reference)
+    : CompAssembly(n, reference), IObjComponent(nullptr), m_gridBase(nullptr),
+      m_minDetId(0), m_maxDetId(0) {
+  init();
+  this->setName(n);
+  setGeometryHandler(new GeometryHandler(this));
+}
+
+bool GridDetector::compareName(const std::string &proposedMatch) {
+  static const boost::regex exp("GridDetector|gridDetector|"
```

Comment:
very minor point, but you could just express the pattern as 
`boost::regex exp("grid_?detector",boost::regex::icase);`